src/emulate.c

```diff
@@ -290,7 +291,16 @@ static uint32_t csr_csrrw(riscv_t *rv, uint32_t csr, uint32_t val)
         out &= FFLAG_MASK;
 #endif
 
-    *c = val;
+    if (c == &rv->csr_satp) {
```

Comment:
> Add comments to address SATP.

I think this check is a bit redundant (I will remove them) because the ppn of root page table (`MASK(22)`)  can be simply obtained during the `mmu_walk`. In addition, the ASID bit could be used to implement TLB. So, store the whole 32-bit value is harmless.